### PR TITLE
LM-30: Update to event is allowed without any changes to event

### PR DESCRIPTION
**Pull request by kkuczkowski-unosq**
Jira issue: # [LM-30](https://unosquarebelfast.atlassian.net/browse/LM-30)
- Added changes to disallow updating events with unchanged event start date & end date

- Made event message (`reason for updating`) a required field (this will need to be reflected in front end)

Adjusted existing unit tests and added new test case providing test coverage for these changes.

Todo:

- [x] raise issue for front end (`reason for updating should be made required field`)

- [x] raise issue for back end (`make booking type (event type) updatable attribute`)

### CODE REVIEW

File: AdminCore.DAL/Models/Event.cs [hunk 1]
```diff
@@ -28,7 +28,7 @@ public class Event
     public DateTime LastModified { get; set; }
 
     [ForeignKey("EventId")]
-    public ICollection<EventMessage> EventMessages { get; set; }
+    public IList<EventMessage> EventMessages { get; set; }
```
Comment (rgillen-uno) [inline AdminCore.DAL/Models/Event.cs:31]: I _think_ this change would be ok. "A navigation property that represents the "many" end of a relationship must return a type that implements ICollection, where T is the type of the object at the other end of the relationship."

IList implements ICollection so I think this should be OK? Just want to make sure this doesn't break the navigation property.

https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/dd468057(v=vs.100)

Comment (kkuczkowski-unosq) [reply AdminCore.DAL/Models/Event.cs:31]: AFAIK this is a non breaking change. For context, I chose to use a more specific `IList` interface as we should be using `List` to take advantage of its orderedness.

File: AdminCore.Services/EventService.cs [hunk 2]
```diff
@@ -146,17 +147,51 @@ public EventDto CreateEvent(EventDateDto dates, EventTypes eventTypes, Employee
...
+    { 
+      var eventToUpdate = GetEventById(dates.EventId);
+      var currentDates = eventToUpdate.EventDates;
+
+      if (EventIsNotUpdatable(eventToUpdate, message))
+      {
+        throw new Exception("Cannot update event, criteria not met");
+      }
+
+      EvaluateEventDates(dates, currentDates, eventToUpdate);
```
Comment (rgillen-uno) [inline AdminCore.Services/EventService.cs:157]: Is there a check for if the if EventType or Half-Day state changes?

Comment (kkuczkowski-unosq) [reply AdminCore.Services/EventService.cs:157]: Good shout with checking `Half-Day` state, forgot about it, EventType state is not supported. In fact, it is only visible in front end, the `UpdateEventViewModel` does not handle `EventType`. I have raised [LM-45](https://unosquarebelfast.atlassian.net/browse/LM-45) to add support for this

**Review by rgillen-uno — APPROVED** [review]

**Review by BrownMark — APPROVED** [review]
Looks good

**Pull request merged**